homeassistant/components/mqtt/fan.py

```diff
@@ -133,7 +183,25 @@ def __init__(self, hass, config, config_entry, discovery_data):
         self._templates = None
         self._optimistic = None
         self._optimistic_oscillation = None
+        self._optimistic_percentage = None
+        self._optimistic_preset_mode = None
         self._optimistic_speed = None
+        if CONF_SPEED_COMMAND_TOPIC in config and CONF_SPEED_LIST in config:
+            if not isinstance(
+                config[CONF_SPEED_LIST], list
+            ) or not speed_list_without_preset_modes(config[CONF_SPEED_LIST]):
+                raise ValueError(
+                    "No valid speeds configured, speeds mode feature disabled"
+                )
+        if (
+            CONF_PRESET_MODE_COMMAND_TOPIC in config
+            and CONF_PRESET_MODE_COMMAND_TOPIC in config
+            and (
+                not config[CONF_PRESET_MODES_LIST]
+                or not isinstance(config[CONF_PRESET_MODES_LIST], list)
```

Comment:
I think check for a list type could be removed since this is checked during schema validation.
A legacy speeds configuration could be ["off"] (no speeds in the list) but does not supply speeds to the list.
I will commit improved code to remove the list check.